### ISSUE TITLE
[java] UnusedAssignment should be able to ignore unused parameters of overriding methods

### ISSUE THREAD

**Issue by Frettman** [pmd/pmd#4281]
**Affects PMD Version:** 6.49.0

**Rule:** [UnusedAssignment](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedassignment)

**Description:**

First of all, I was kind of surprised that this check also looks at parameters, as that's not quite obvious from the description and their initial value isn't assigned in the traditional sense. A parameter assignment typically means overriding its initial value (to me anyway).
Anyway, I took this rule to mean: _The value is never used, so do you even need the variable, field, etc.?_ Applying that to method parameters: _That parameter is never used. If you don't need, maybe remove it from the method._ But when implementing or overriding methods, that's typically not an option. As such, I would like an option to suppress the violation in that case.

Unused fields can always be removed; unused parameters in overriding methods only potentially. And I'm mainly interested in checks that one (pretty much) always should act upon.

**Running PMD through:** Maven and Eclipse Plug-in

**Comment by Frettman** [comment]
You are right that an unused value could also mean that the value still needs to be used. But that's something that should also come up during testing, if not before. So for me personally, detecting where something unused can be removed is the added benefit of this check that I haven't covered otherwise. But of course I don't expect that to be true for everyone.
Anyway, a flag to optionally ignore parameters of overriding and implementing methods would be much appreciated. By the way, the Eclipse compiler also has an option like that, though that doesn't really help during a build.

**Comment by jsotuyod** [comment]
Well, the rule's doc claims that this rule subsume https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedformalparameter yet, that rule ignores overridden methods automatically (as well as serialization methods), and I think doing both things are very reasonable.

I don't even see value in the property to be honest. If it's in a library there is no point in getting reported as it's outside the analyzed project's scope. Even if it's not, if this is an interface, it is most probably than notbe designed to cover much broader set of use cases (ie: an init listener may provide initial values, but an implementation may not necessarily care about them).

Serialization methods are even worst, as they are not even overrides.

Moreover, UnusedFormalParameter would even ignore non-private methods by default (under a property) in case you wanted to keep api compatibility.

Overall, UnusedFormalParameter is a significantly superior rule, which is natural given how much time it has have to mature and improve on feedback.

Now, this duplicity is a net negative for PMD. It doubles mainting effort, confuses users and lead to duplicate reports or inconsistencies (one rule doesn't *really* replace the other yet both may flag the same thing).

There are two ways to remove it:

1. We deprecate UnusedFormalParameter and UnusedLocalVariable while ensuring this rule really is equivalent.
2. We remove the support to report unused vars / formal params from this rule and let all rules complement each other.

On this particular case, I lean heavily on option 2 following the principle of least surprise. If I want to fish unused variables / params, the list place I would expect it to be is under the UnusedAssignment rule. Users should not read the full docs and all properties' detail to find that. Same wise, having

**Comment by jsotuyod** [comment]
UnusedAssignment

**Comment by oowekyala** [comment]
I still don't agree with this assessment. In the following code
```java
void doSomething(String foo, String bar) {
    foo = bar.trim();
    System.out.println(foo + bar);
}
```
the user is using both variables, yet the initial value of the parameter `foo` is ignored. This is probably a mistake and the user meant to write `foo.trim()`. This is a case that UnusedFormalParameter doesn't check at all, and which is exactly what UnusedAssignment is meant to track (a value that is given a name and not used).

UnusedAssignment already does not report unused variables, to avoid overlap with UnusedVariable and UnusedFormalParameter. This is controlled by a property `reportUnusedVariables` which is off by default. This sounds like your option 2 tbh, as the rules coexist and complement each other using this flag.

I think this is based on a misunderstanding:

> Anyway, I took this rule to mean: The value is never used, so do you even need the variable, field, etc.? Applying that to method parameters: **That parameter is never used. If you don't need, maybe remove it from the method.** But when implementing or overriding methods, that's typically not an option. As such, I would like an option to suppress the violation in that case.

This is not what the rule means. When you keep the defaults of the rule, you don't see reports of unused formal parameters so you cannot remove the parameter. What the violations of this rule mean is **A named value is unused. Either 1. use it, 2. remove the assignment (or parameter), or 3. rename the variable to `unused` or so to make it clear this is intended.** Solution 2 is not always possible given parameters of public/overridden methods, or exception parameters for instance. That's what solution 3 is for. These solutions are equally possible in overridden methods so I don't see why we should special case them and remove the ability to flag examples like the above in non-private methods.

> On this particular case, I lean heavily on option 2 following the principle of least surprise. If I want to find unused variables / params, the last place I would expect it to be is under the UnusedAssignment rule. Users should not read the full docs and all properties' detail to find that. Samewise, having UnusedAssignment flag something that is not even an assignment is contradictory, regardless of properties.

I believe given the defaults of UnusedAssignment this is already implemented. Maybe the rule can be renamed UnusedNamedValue to be less surprising, but the implementation behaves in a useful way. So I don't think we should cripple the implementation to fit the name, but the reverse, if at all.

**Comment by jsotuyod** [comment]
> I believe given the defaults of UnusedAssignment this is already implemented. Maybe the rule can be renamed UnusedNamedValue to be less surprising, but the implementation behaves in a useful way. So I don't think we should cripple the implementation to fit the name, but the reverse, if at all.

I don't fully agree. Deprecating the property and eventually removing it is not really crippling the rule, but avoiding the overlap between different and complementing rules. The default behavior for the rule is still useful on it's own, as it detects different things as you said, although the doc and examples can probably make these clearer.

Doing the reverse means having to support and update multiple rules in sync (again, overrides, serialization method, etc.) which is very error prone (ie: remembering any FP / FN on one rule needs to be checked on the other) and just overhead, as it creates no additional value to users, which can still achieve the same one way or another.

**Comment by oowekyala** [comment]
The reason the property is there is that the implementation of UnusedAssignment naturally subsumes that of the other unused rules. In other words, a completely unused variable is a special case of unused assignment. So removing the property without enabling these overlapping violations will not simplify the rule. It will always need to somehow explicitly exclude these violations to avoid the overlap. 

Actually I just realized we are talking about different things here which might be why we can't agree
- when I meant crippling the rule I meant removing its ability to report on parameters altogether, which you seemed to be suggesting, and which IMHO was implied by the fact that the OP didn't turn on the `reportUnusedVariable` property. It appears from your last message that you mean just removing the property and make it assume its default value? 
- when I meant "doing the reverse" I just meant renaming the rule so that its name fit its implementation. Apparently you mean making the behavior of the rule when `reportUnusedVariable` is true match the behavior of UnusedFormalParameter and UnusedVariable, by excluding a bunch of things like overridden methods and such.

If I understood correctly what you mean then we would just remove the property and make it assume its default value. In my opinion that's fine. It may be useful to remove the property, because probably no one will turn it on. I just don't want UnusedAssignment to accumulate a bunch of special cases like the other rules. Unused assignments are a problem also in overridden, in default methods, etc. So this ticket should be a won't fix and maybe we should move to another issue.

> On this I agree. Maybe UnusedVariableValue would be even better?

I think this is more ambiguous than UnusedNamedValue, because it could be read as "unused value that is variable" instead of "unused value of a variable"...

**Comment by oowekyala** [comment]
> Yeah, maybe you are right… having said that, both names could be confusing, as it only works on local scope vars, not all "named values", so we still have [UnusedPrivateField](https://docs.pmd-code.org/latest/pmd_rules_java_bestpractices.html#unusedprivatefield)…

It does work on fields though. For instance: https://github.com/pmd/pmd/blob/master/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/UnusedAssignment.xml#L3061-L3062

The error messages are already pretty well differentiated. You would get something like eg https://github.com/pmd/pmd/blob/d4b99bb047c832f4b0b7153dc91b23f6b8d19565/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/UnusedAssignment.xml#L1878

The fix in general always requires some care because the expression may have some side effects... I don't see what's wrong with the current messages tbh

**Comment by jsotuyod** [comment]
Awesome!

**Comment by Frettman** [comment]
I created this ticket assuming that `UnusedAssignment` would one day replace `UnusedLocalVariable` and `UnusedFormalParameter` (with `reportUnusedVariables` = true), based on that one sentence from the documentation: `The rule subsumes UnusedLocalVariable, and UnusedFormalParameter.` But neither are deprecated even now, so this just looks like a misunderstanding or miscommunication. Especially after this discussion.
I will just ignore the `reportUnusedVariables` option and the rule should coexist nicely with the other `Unused*` rules. As such, the original point of this ticket is moot. Clarifying the documentation or even removing the `reportUnusedVariables` flag seems like a good idea, but yeah, that should probably be a different ticket. So feel free to close this.
